Add developer's docs on XBlock Asides (#1443)
openedx/docs.openedx.org: source/developers/how-tos/add-an-xblock-aside.rst

when you want a single, installable Python package that decorates the
views of XBlocks in your platform.

For background on what asides are and when to use them, read
:ref:`About XBlock Asides`. For a complete API reference, see
:ref:`XBlock Asides Reference`.

.. contents:: Contents
   :local:
   :depth: 1

Prerequisites
*************

Before you start, make sure you have:

* A working Open edX development environment in which you can install a
  Python package and restart the LMS and Studio services. Tutor devstack
  is the recommended environment.
* The installed Python version used by the target Open edX release.
* Familiarity with writing a basic XBlock view that returns a
  :class:`~web_fragments.fragment.Fragment`. If you have never written
  one, complete :ref:`XBlock Aside Quickstart` first.

This recipe builds a feedback-badge Aside that adds a "Report an issue"
link to Problem and Video blocks, with a course-author setting to enable
or disable it per block. Substitute your own block types and behavior as
needed.

Step 1: Scaffold a Python package
*********************************

Create a new directory for the Aside package, with the layout below:

> 
> .. code-block:: text
> 
>    feedback_badge_aside/
>    ├── pyproject.toml
>    ├── feedback_badge_aside/
>    │   ├── __init__.py
>    │   └── Aside.py
>    └── README.rst

Reviewer: ```suggestion

.. code-block:: text

   feedback_badge_aside/
   ├── pyproject.toml
   ├── feedback_badge_aside/
   │   ├── __init__.py
   │   ├── Aside.py
   │   └── static/
   │       └── js/
   │           └── studio.js
   └── README.rst
```

Reviewer: This adds a JS file in the structure.

   dependencies = [
       "XBlock",
       "web-fragments",
   ]

   [build-system]
   requires = ["setuptools>=61.0"]
   build-backend = "setuptools.build_meta"

Step 2: Define the Aside class
******************************

In ``feedback_badge_aside/Aside.py``, define a subclass of
:class:`~xblock.core.XBlockAside`.

.. code-block:: python

   from xblock.core import XBlockAside

   class FeedbackBadgeAside(XBlockAside):
       """Adds a feedback link to learner-facing views of supported blocks."""

The class name does not need to match the entry point name, but keeping
them consistent makes debugging easier.

Step 3: Declare fields for course-author control
************************************************

Add a Boolean field that course authors can toggle to enable or disable
the Aside on a per-block basis. Scope the field to ``Scope.settings``,
which means the value is stored per block and travels with the course in
OLX export and import.

.. code-block:: python

   from xblock.core import XBlockAside
   from xblock.fields import Boolean, Scope

   class FeedbackBadgeAside(XBlockAside):
       """Adds a feedback link to learner-facing views of supported blocks."""

>        enabled = Boolean(

Reviewer: Having the same field name across multiple asides can create conflicts, so it should be unique.
```suggestion
       is_feedback_aside_enabled = Boolean(
```

   class FeedbackBadgeAside(XBlockAside):
       """Adds a feedback link to learner-facing views of supported blocks."""

       enabled = Boolean(
           display_name="Show feedback link",
           default=True,
           scope=Scope.settings,
           help="Whether to show a 'Report an issue' link on this block.",
       )

       @XBlockAside.aside_for("student_view")
       def student_view_aside(self, block, context=None):
           """Render the feedback link for the learner view."""
           if not self.enabled:
               return Fragment("")

           block_id = block.scope_ids.usage_id.block_id
           html = (
               f'<a class="feedback-badge" '
               f'href="/feedback?block={block_id}">Report an issue</a>'
           )
           return Fragment(html)

       @XBlockAside.aside_for("studio_view")
       def studio_view_aside(self, block, context=None):
           """Render the author-side toggle UI in Studio."""
           checked = "checked" if self.enabled else ""
           html = (
               f'<label><input type="checkbox" {checked} '
               f'class="feedback-badge-toggle"> Show feedback link</label>'
           )

>            return Fragment(html)

Reviewer: Changing the state of this checkbox won't save the setting. We will need to implement a handler and write custom JS to call the handler when the state of this field changes to save the setting. I have mentioned this in my other comment in about_xblock_asides.rst

Author: @asadali145 would you be able to provide some inline corrections in this diff, or some alternative text, to replace this?

>        @XBlockAside.aside_for("studio_view")
>        def studio_view_aside(self, block, context=None):
>            """Render the author-side toggle UI in Studio."""
>            checked = "checked" if self.enabled else ""
>            html = (
>                f'<label><input type="checkbox" {checked} '
>                f'class="feedback-badge-toggle"> Show feedback link</label>'
>            )
>            return Fragment(html)

Reviewer: ```suggestion
    @XBlockAside.aside_for("author_view")
    def author_view_aside(self, block, context=None):
        """Render the author-side toggle UI."""

        checked = "checked" if self.enabled else ""

        html = (
            f'<label>'
            f'<input type="checkbox" '
            f'class="feedback-badge-toggle" {checked}> '
            f'Show feedback link'
            f'</label>'
        )

        fragment = Fragment(html)
        fragment.add_javascript_url(
            self.runtime.local_resource_url(self, "static/js/studio.js")
        )
        fragment.initialize_js("FeedbackBadgeAside")
        return fragment

    @XBlock.handler
    def update_config(self, request, suffix=""):
        """Update the aside configuration."""

        try:
            posted_data = request.json
        except ValueError:
            return Response(
                "Invalid request body",
                status=api_status.HTTP_400_BAD_REQUEST,
            )

        self.enabled = posted_data.get("is_enabled", True)
        return Response()
```

Reviewer: This changes `studio_view` to `author_view`, adds the JS file in the fragment, and adds an aside handler to be able to save the aside field data from the authoring unit view. If we do not create this handler and custom JS, we cannot change the Aside field like enable/disable.

For production code, render templates from files with the runtime's
template service rather than building HTML strings inline. The strings
above keep the example readable.

> 

Reviewer: ```
Step 5: Add custom JS to save the aside field data

The accompanying `studio.js` listens for changes to the checkbox and sends the updated configuration to the Aside's handler. Add the below to the studio.js.
```
```
.. code-block:: javascript

   (function ($) {
       'use strict';

       function FeedbackBadgeAside(runtime, element) {
           var $element = $(element);
           var updateInProgress = false;

           // Studio runtime is required to generate handler URLs.
           var studioRuntime = new window.StudioRuntime.v1();

           $element.find('.feedback-badge-toggle').on('change', function (e) {
               if (updateInProgress) {
                   return;
               }

               updateInProgress = true;

               e.preventDefault();

               runtime.notify('save', {
                   state: 'start',
                   element: element,
                   message: gettext('Saving configuration')
               });

               $.ajax({
                   type: 'POST',
                   url: studioRuntime.handlerUrl(element, 'update_config'),
                   data: JSON.stringify({
                       is_enabled: this.checked
                   }),
                   dataType: 'json',
                   contentType: 'application/json; charset=utf-8'
               }).done(function () {
                    try {
                        window.parent.postMessage({
                            type: 'saveEditedXBlockData',
                            message: 'Sends a message when the xblock data is saved',
                            payload: {}
                        }, document.referrer);
                    } catch (e) {
                        console.error(e);
                    }
                }).always(function () {
                   runtime.notify('save', {
                       state: 'end',
                       element: element
                   });

                   updateInProgress = false;
               });
           });
       }

       function initializeFeedbackBadgeAside(runtime, element) {
           return new FeedbackBadgeAside(runtime, element);
       }

       window.FeedbackBadgeAsideInit = initializeFeedbackBadgeAside;
   }($));

Reviewer: This is to add a new step to add the JS code. JS code basically calls the aside handler to added in the aside.py when the checkbox changes on the frontend.

* Course-level settings retrieved through a runtime service.

If your filter consults course settings or feature flags, guard against
the import and export paths where these may not be available; see
:ref:`About XBlock Asides` for the relevant limitations.

.. _register entry point:

Step 6: Register the Aside as an entry point
********************************************

In ``pyproject.toml``, add an entry point in the ``xblock_asides.v1``
group. The entry point name on the left side of the equals sign becomes
the Aside's type name and is used as the XML tag during OLX
serialization.

.. code-block:: toml

   [project.entry-points."xblock_asides.v1"]
   feedback_badge = "feedback_badge_aside.Aside:FeedbackBadgeAside"

Choose a type name that is unlikely to collide with other asides on the
same deployment. Treat the name as a stable public identifier; renaming
it later breaks OLX round-trips of any course that has used the Aside.

Step 7: Install the package and restart services
************************************************

Install the package into the LMS and Studio Python environments. With
Tutor:

.. code-block:: bash

   tutor mounts add /path/to/feedback_badge_aside

>    tutor dev launch

Reviewer: I couldn't test this with mounts. I'll give it another go later today.

registration.

Open the LMS Django admin and create a new configuration revision:

.. code-block:: text

   http://<your-lms-host>/admin/lms_xblock/xblockasidesconfig/

Click :guilabel:`Add`, check :guilabel:`Enabled`, and save. The model is
a ``ConfigurationModel`` (revision-based), so each save creates a new
revision and the most recent enabled revision is treated as current.

The same form has a ``Disabled blocks`` field, a space-separated list of
block types on which asides will **never** render in the LMS. The
default value is ``about course_info static_tab``. If your Aside should
apply to one of these block types, remove that type from the list.

There is no per-course allowlist and no per-Aside-type allowlist. Once
``XBlockAsidesConfig`` is enabled and your Aside's host block type is
not in ``disabled_blocks``, the runtime offers your Aside to every
matching block in every course. Per-Aside filtering happens through
your Aside's own ``should_apply_to_block`` classmethod, which you wrote
in Step 5.

The Studio runtime does not consult this configuration. Asides render
in Studio author views independently of ``XBlockAsidesConfig``, as soon

> as they are installed and registered.

Reviewer: As mentioned in Quickstart, Studio has its own model http://studio.local.openedx.io:8001/admin/xblock_config/studioconfig/. We need to enable it to enable the aside author view.

Author: @asadali145 I don't think I quite follow, would you be able to provide some inline corrections in this diff, or some alternative text, to replace this?

Reviewer: ```suggestion
The Studio also gates Asides rendering on a global Django
configuration model, ``StudioConfig``. This is similar to 
LMS ``XBlockAsidesConfig`` and needs to be enabled to 
render the Asides in studio.

Open the Studio Django admin at:

.. code-block:: text

   http://<your-studio-host>/admin/xblock_config/studioconfig/

```

.. _Add an XBlock Aside:

####################
Add an XBlock Aside
####################

.. tags:: developer, how-to

> Add an XBlock aside to attach behavior, UI, or stored data to one or more

Reviewer 2: ```suggestion
Add an XBlock Aside to attach behavior, UI, or stored data to one or more
```

@sarina: I'm on the fence about whether to capitalize "Aside" in these docs. I've generally been doing so in chat. What do you think?

* Course-level settings retrieved through a runtime service.

If your filter consults course settings or feature flags, guard against
the import and export paths where these may not be available; see
:ref:`About XBlock Asides` for the relevant limitations.

.. _register entry point:

Step 6: Register the aside as an entry point
********************************************

In ``pyproject.toml``, add an entry point in the ``xblock_asides.v1``
group. The entry point name on the left side of the equals sign becomes
the aside's type name and is used as the XML tag during OLX
serialization.

.. code-block:: toml

   [project.entry-points."xblock_asides.v1"]
   feedback_badge = "feedback_badge_aside.aside:FeedbackBadgeAside"

Choose a type name that is unlikely to collide with other asides on the
same deployment. Treat the name as a stable public identifier; renaming
it later breaks OLX round-trips of any course that has used the aside.

Step 7: Install the package and restart services
************************************************

Install the package into the LMS and Studio Python environments. With
Tutor:

.. code-block:: bash

   tutor mounts add /path/to/feedback_badge_aside

>    tutor dev launch

Reviewer 2: Notes to self:

1. I don't think auto-mounting would work for this (i.e. we need the more verbose form)?
2. This is assuming people are running Tutor dev and not local.

Reviewer: FYI that ol_openedx_chat Aside didn't work for me like this.

Author: @asadali145 would you be able to provide some inline corrections in this diff, or some alternative text, to replace this?

Or, if you are managing the environment manually:

.. code-block:: bash

   pip install -e /path/to/feedback_badge_aside

> 

Reviewer 2: ```suggestion
```

I feel like this can just get folks in trouble unless they really know what they're doing (in which case, they already know they can do this).

Reviewer 3: agree

After installing, restart both the LMS and Studio. Asides are discovered
at process start, so newly installed asides do not appear until the
services restart.

> 

Reviewer 2: ```suggestion
```

If we're having them do `tutor dev launch`, it's going to rebuild everything and restart anyhow.

registration.

Open the LMS Django admin and create a new configuration revision:

.. code-block:: text

   http://<your-lms-host>/admin/lms_xblock/xblockasidesconfig/

Click :guilabel:`Add`, check :guilabel:`Enabled`, and save. The model is
a ``ConfigurationModel`` (revision-based), so each save creates a new
revision and the most recent enabled revision is treated as current.

The same form has a ``Disabled blocks`` field, a space-separated list of
block types on which asides will **never** render in the LMS. The
default value is ``about course_info static_tab``. If your aside should
apply to one of these block types, remove that type from the list.

There is no per-course allowlist and no per-aside-type allowlist. Once
``XBlockAsidesConfig`` is enabled and your aside's host block type is
not in ``disabled_blocks``, the runtime offers your aside to every
matching block in every course. Per-aside filtering happens through
your aside's own ``should_apply_to_block`` classmethod, which you wrote
in Step 5.

The Studio runtime does not consult this configuration. Asides render
in Studio author views independently of ``XBlockAsidesConfig``, as soon

> as they are installed and registered.

Reviewer 2: Note to self: Verify this and file a bug if necessary.

Reviewer: Not a bug, studio has its own model to gate asides.

Step 9: Verify the aside is rendering
*************************************

Open a course that contains a Problem or Video block, view it as a
learner, and confirm the feedback link appears at the bottom of the
block. To verify the author-side UI, open the same block in Studio and
confirm the toggle appears in the studio view.

If the aside does not appear:

#. Check that the entry point is registered. Run:

   .. code-block:: python

      from xblock.core import XBlockAside
      print(list(XBlockAside.load_classes()))

   in a Django shell. Your aside's type name should be in the list.

#. Check that ``should_apply_to_block`` returns ``True`` for the block
   you are testing.

> #. Check that the aside is allowlisted for the course.

Reviewer 2: ```suggestion
```
